File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/UsersMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.conversations.CallbackListener
import com.twilio.util.ErrorInfo
import com.twilio.conversations.User
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import android.util.Log

object UsersMethods {
    fun getChannelUserDescriptors(call: MethodCall, result: MethodChannel.Result) {
        result.success(Mapper.usersToMap(TwilioConversationsPlugin.chatClient?.subscribedUsers!!))
    }

    fun getUserDescriptor(call: MethodCall, result: MethodChannel.Result) {
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        TwilioConversationsPlugin.chatClient?.getAndSubscribeUser(identity, object : CallbackListener<User> {
            override fun onSuccess(userDescriptor: User) {
                Log.d("TwilioInfo", "UsersMethods.getUserDescriptor => onSuccess")
                result.success(Mapper.userDescriptorToMap(userDescriptor))
            }

            override fun onError(errorInfo: ErrorInfo) {
                Log.d("TwilioInfo", "UsersMethods.getUserDescriptor => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getAndSubscribeUser(call: MethodCall, result: MethodChannel.Result) {
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        TwilioConversationsPlugin.chatClient?.getAndSubscribeUser(identity, object : CallbackListener<User> {
            override fun onSuccess(user: User) {
                Log.d("TwilioInfo", "UsersMethods.getAndSubscribeUser => onSuccess")
                result.success(Mapper.userToMap(user))
            }

            override fun onError(errorInfo: ErrorInfo) {
                Log.d("TwilioInfo", "UsersMethods.getAndSubscribeUser => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }
}

File: flutter_twilio_conversations/ios/Classes/Methods/UsersMethods.swift
import Flutter
import Foundation
import TwilioConversationsClient

public class UsersMethods {
    public static func getChannelUserDescriptors(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.conversation(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHConversation?) in
            if result.isSuccessful, let channelRes = channel as TCHConversation? {
                // FIXME not sure to get users per conversation
                SwiftTwilioConversationsPlugin.debug("UsersMethods.getChannelUserDescriptors => onError: NOT_IMPLEMENTED")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving user descriptors for channel: NOT_IMPLEMENTED", details: nil))
            } else {
                SwiftTwilioConversationsPlugin.debug("UsersMethods.getChannelUserDescriptors => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel: \(channelSid)", details: nil))
            }
        })
    }

    public static func getUserDescriptor(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.subscribedUser(withIdentity: identity, completion: { (result: TCHResult, user: TCHUser?) in
            if result.isSuccessful {
                if let userDesc = user as TCHUser? {
                    SwiftTwilioConversationsPlugin.debug("UsersMethods.getUserDescriptor => onSuccess")
                    flutterResult(Mapper.userDescriptorToDict(userDesc))
                } else {
                    flutterResult(FlutterError(code: "ERROR", message: "Could not resolve userDescriptor for identity: \(identity)", details: nil))
                }
            } else {
                SwiftTwilioConversationsPlugin.debug("UsersMethods.getUserDescriptor => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving userDescriptor for identity: \(identity)", details: nil))
            }
        })
    }

    public static func getAndSubscribeUser(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.subscribedUser(withIdentity: identity, completion: { (result: TCHResult, user: TCHUser?) in
            if result.isSuccessful, let user = user {
                flutterResult(Mapper.userToDict(user))
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error subscribing to user with identity: \(identity)", details: nil))
            }
        })
    }
}
